### ISSUE TITLE
PerlIO_tmpfile: Fall back to cwd if we have no /tmp or TMPDIR

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13432]
Migrated from [rt.perl.org#120591](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120591) (status was 'resolved')

Searchable as RT120591$

**Comment by p5pRT** [comment]
### From @Hugmeir

#### Created by @Hugmeir



With this\, open\($fh\, undef\) will now work on systems without
a /tmp \(or equivalent\) where TMPDIR is not set\.
The OS that prompted this was Android\, which has no canonical temporary
directory;
the only place it has mounted as a tmpfs is /mnt/asec\, but that requires
root\-level
privileges\.

\-\-\-
 perlio\.c | 8 \+\+\+\+\+\+\+\+
 1 file changed\, 8 insertions\(\+\)

<details><summary>Inline Patch</summary>

```diff
diff --git a/perlio.c b/perlio.c
index 60b6a59..c2ce8f6 100644
--- a/perlio.c
+++ b/perlio.c
@@ -5039,6 +5039,14 @@ PerlIO_tmpfile(void)
      /* else we try /tmp */
      fd = mkstemp(tempname);
      }
+     if (fd < 0) {
+         /* Try cwd */
+         if ( sv )
+             SvREFCNT_dec(sv);
+         sv = newSVpvs(".");
+         sv_catpv(sv, tempname + 4);
+         fd = mkstemp(SvPVX(sv));
+     }
      if (fd >= 0) {
       f = PerlIO_fdopen(fd, "w+");
       if (f)
-- 
1.8.3.2
```
</details>


<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.16.2:

Configured by hugmeir at Tue Nov 20 17:20:00 ART 2012.

Summary of my perl5 (revision 5 version 16 subversion 2) configuration:

  Platform:
    osname=linux, osvers=3.5.0-18-generic,
archname=x86_64-linux-thread-multi
    uname='linux naw 3.5.0-18-generic #29-ubuntu smp fri oct 19 10:26:51
utc 2012 x86_64 x86_64 x86_64 gnulinux '
    config_args='-de
-Dprefix=/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2 -DDEBUGGING
-Dusethreads -Doptimize=-g -O0 -ggdb3 -Uversiononly -Accflags=-Wall -Wextra
-Aeval:scriptdir=/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -Wall -Wextra -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g -O0 -ggdb3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -Wall -Wextra -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.7.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib
/usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.15'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -O0 -ggdb3 -L/usr/local/lib
-fstack-protector'

Locally applied patches:



@INC for perl 5.16.2:
    /home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/lib/perl5/x86_64-linux-thread-multi
    /home/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5

/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/site_perl/5.16.2/x86_64-linux-thread-multi
    /home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/site_perl/5.16.2

/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/5.16.2/x86_64-linux-thread-multi
    /home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/5.16.2
    .


Environment for perl 5.16.2:
    HOME=/home/hugmeir
    LANG=en_US.UTF-8
    LANGUAGE=en_US
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/bin:/home/hugmeir/perl5/perlbrew/bin:/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin:/usr/lib/lightdm/lightdm:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
    PERL5LIB=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5
    PERLBREW_BASHRC_VERSION=0.66
    PERLBREW_HOME=/home/hugmeir/.perlbrew
    PERLBREW_LIB=all
    PERLBREW_MANPATH=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/man:/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/man
    PERLBREW_PATH=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
/bin:/home/hugmeir/perl5/perlbrew/bin:/home/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin
    PERLBREW_PERL=perl-5.16.2
    PERLBREW_ROOT=/home/hugmeir/perl5/perlbrew
    PERLBREW_VERSION=0.66
    PERL_BADLANG (unset)
    PERL_LOCAL_LIB_ROOT=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
    PERL_MB_OPT=--install_base /home/hugmeir/.perlbrew/libs/perl-5.16.2@all
    PERL_MM_OPT=INSTALL_BASE=/home/hugmeir/.perlbrew/libs/perl-5.16.2@all
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @Hugmeir

<details><summary>0001-perlio.c-PerlIO_tmpfile-Fall-back-to-cwd-if-we-have-.patch</summary>

```diff
From 81b7b6b07c88dbba2dcb85c30c5a84837c925cc5 Mon Sep 17 00:00:00 2001
From: Brian Fraser <fraserbn@gmail.com>
Date: Fri, 26 Apr 2013 02:24:14 -0300
Subject: [PATCH 2/9] perlio.c, PerlIO_tmpfile: Fall back to cwd if we have no
 /tmp or TMPDIR

With this, open($fh, undef) will now work on systems without
a /tmp (or equivalent) where TMPDIR is not set.
---
 perlio.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/perlio.c b/perlio.c
index 60b6a59..c2ce8f6 100644
--- a/perlio.c
+++ b/perlio.c
@@ -5039,6 +5039,14 @@ PerlIO_tmpfile(void)
 	 /* else we try /tmp */
 	 fd = mkstemp(tempname);
      }
+     if (fd < 0) {
+         /* Try cwd */
+         if ( sv )
+             SvREFCNT_dec(sv);
+         sv = newSVpvs(".");
+         sv_catpv(sv, tempname + 4);
+         fd = mkstemp(SvPVX(sv));
+     }
      if (fd >= 0) {
 	  f = PerlIO_fdopen(fd, "w+");
 	  if (f)
-- 
1.8.3.2


```
</details>

**Comment by p5pRT** [comment]
### From @bulk88

On Tue Nov 19 19&#8203;:22&#8203;:13 2013\, Hugmeir wrote&#8203;:
> This is a bug report for perl from fraserbn@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> With this\, open\($fh\, undef\) will now work on systems without
> a /tmp \(or equivalent\) where TMPDIR is not set\.
> The OS that prompted this was Android\, which has no canonical temporary
> directory;
> the only place it has mounted as a tmpfs is /mnt/asec\, but that requires
> root\-level
> privileges\.
> 
> \-\-\-
>  perlio\.c | 8 \+\+\+\+\+\+\+\+
>  1 file changed\, 8 insertions\(\+\)
> 
> diff \-\-git a/perlio\.c b/perlio\.c
> index 60b6a59\.\.c2ce8f6 100644
> \-\-\- a/perlio\.c
> \+\+\+ b/perlio\.c
> @&#8203;@&#8203; \-5039\,6 \+5039\,14 @&#8203;@&#8203; PerlIO\_tmpfile\(void\)
>       /\* else we try /tmp \*/
>       fd = mkstemp\(tempname\);
>       \}


why are you testing fd \< 0 when the block directly above with "mkstemp" is fd \< 0 conditional?

> \+     if \(fd \< 0\) \{
> \+         /\* Try cwd \*/

SV is already NULL due a NULL assignment earlier remove the dec\, also SvREFCNT\_dec will safely take a NULL SV\, use different variant of SvREFCNT\_dec
> \+         if \( sv \)
> \+             SvREFCNT\_dec\(sv\);
> \+         sv = newSVpvs\("\."\);
> \+         sv\_catpv\(sv\, tempname \+ 4\);
> \+         fd = mkstemp\(SvPVX\(sv\)\);
> \+     \}
>       if \(fd >= 0\) \{
>        f = PerlIO\_fdopen\(fd\, "w\+"\);
>        if \(f\)

actually\, something looks funny in the existing code


---
&nbsp;     SV \* sv = NULL;
&nbsp;     /\*
&nbsp;      \* I have no idea how portable mkstemp\(\) is \.\.\. NI\-S
&nbsp;      \*/
&nbsp;     if \(tmpdir && \*tmpdir\) \{
&nbsp;     /\* if TMPDIR is set and not empty\, we try that first \*/
&nbsp;     sv = newSVpv\(tmpdir\, 0\);
&nbsp;     sv\_catpv\(sv\, tempname \+ 4\);
&nbsp;     fd = mkstemp\(SvPVX\(sv\)\);
&nbsp;     \}
&nbsp;     if \(fd \< 0\) \{
&nbsp;     sv = NULL; \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\# didn't "sv = newSVpv\(tmpdir\, 0\);" just leak?
&nbsp;     /\* else we try /tmp \*/
&nbsp;     fd = mkstemp\(tempname\);
&nbsp;     \}


---
\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @bulk88

On Wed Nov 20 23&#8203;:06&#8203;:38 2013\, bulk88 wrote&#8203;:
> > diff \-\-git a/perlio\.c b/perlio\.c
> > index 60b6a59\.\.c2ce8f6 100644
> > \-\-\- a/perlio\.c
> > \+\+\+ b/perlio\.c
> > @&#8203;@&#8203; \-5039\,6 \+5039\,14 @&#8203;@&#8203; PerlIO\_tmpfile\(void\)
> >       /\* else we try /tmp \*/
> >       fd = mkstemp\(tempname\);
> >       \}
> 
> 
> why are you testing fd \< 0 when the block directly above with
> "mkstemp" is fd \< 0 conditional?
> 
> > \+     if \(fd \< 0\) \{
> > \+         /\* Try cwd \*/

I'm blind\, nvm on that question\, I see the fd assignment\.
\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @Hugmeir

On Thu\, Nov 21\, 2013 at 4&#8203;:06 AM\, bulk88 via RT \<perlbug\-followup@&#8203;perl\.org>wrote&#8203;:

> On Tue Nov 19 19&#8203;:22&#8203;:13 2013\, Hugmeir wrote&#8203;:
> > This is a bug report for perl from fraserbn@&#8203;gmail\.com\,
> > generated with the help of perlbug 1\.39 running under perl 5\.16\.2\.
> >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> >
> > With this\, open\($fh\, undef\) will now work on systems without
> > a /tmp \(or equivalent\) where TMPDIR is not set\.
> > The OS that prompted this was Android\, which has no canonical temporary
> > directory;
> > the only place it has mounted as a tmpfs is /mnt/asec\, but that requires
> > root\-level
> > privileges\.
> >
> > \-\-\-
> >  perlio\.c | 8 \+\+\+\+\+\+\+\+
> >  1 file changed\, 8 insertions\(\+\)
> >
> > diff \-\-git a/perlio\.c b/perlio\.c
> > index 60b6a59\.\.c2ce8f6 100644
> > \-\-\- a/perlio\.c
> > \+\+\+ b/perlio\.c
> > @&#8203;@&#8203; \-5039\,6 \+5039\,14 @&#8203;@&#8203; PerlIO\_tmpfile\(void\)
> >       /\* else we try /tmp \*/
> >       fd = mkstemp\(tempname\);
> >       \}
>
>
> why are you testing fd \< 0 when the block directly above with "mkstemp" is
> fd \< 0 conditional?
>
> > \+     if \(fd \< 0\) \{
> > \+         /\* Try cwd \*/
>
> SV is already NULL due a NULL assignment earlier remove the dec\, also
> SvREFCNT\_dec will safely take a NULL SV\, use different variant of
> SvREFCNT\_dec
> > \+         if \( sv \)
> > \+             SvREFCNT\_dec\(sv\);
> > \+         sv = newSVpvs\("\."\);
> > \+         sv\_catpv\(sv\, tempname \+ 4\);
> > \+         fd = mkstemp\(SvPVX\(sv\)\);
> > \+     \}
> >       if \(fd >= 0\) \{
> >        f = PerlIO\_fdopen\(fd\, "w\+"\);
> >        if \(f\)
>
> actually\, something looks funny in the existing code
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>      SV \* sv = NULL;
>      /\*
>       \* I have no idea how portable mkstemp\(\) is \.\.\. NI\-S
>       \*/
>      if \(tmpdir && \*tmpdir\) \{
>          /\* if TMPDIR is set and not empty\, we try that first \*/
>          sv = newSVpv\(tmpdir\, 0\);
>          sv\_catpv\(sv\, tempname \+ 4\);
>          fd = mkstemp\(SvPVX\(sv\)\);
>      \}
>      if \(fd \< 0\) \{
>          sv = NULL; \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\# didn't "sv = newSVpv\(tmpdir\, 0\);"
> just leak?
>          /\* else we try /tmp \*/
>          fd = mkstemp\(tempname\);
>      \}
>

Yeah\, that code was leaking\. Well spotted\! Fixed
in e40f8e806ef39ca48d2e94f25d2b6c5f63b4efda
Thanks for the feedback; I've merged an updated version of the patch
as b7561fc9bc645b8c7e0e9ebbb29349113cb716c1\, so I'll closing this ticket\.

**Comment by p5pRT** [comment]
### From @Hugmeir

Fixed in b7561fc9bc645b8c7e0e9ebbb29349113cb716c1\, closing\.

\-\-hugmeir

**Comment by p5pRT** [comment]
@Hugmeir - Status changed from 'open' to 'resolved'
